packages/hydrogen/src/product/useOptimisticProduct.ts

```diff
@@ -0,0 +1,104 @@
+  selectedVariant?: PartialDeep<ProductVariant>;
+};
+
+type ProductWithVariants = {
+  product: {
+    variants: {nodes: Array<PartialDeep<ProductVariant>>};
+  };
+};
+
+type OptimisticProductVariants =
+  | Array<PartialDeep<ProductVariant>>
+  | Promise<Array<PartialDeep<ProductVariant>>>
+  | ProductWithVariants
+  | Promise<ProductWithVariants>;
+
+/**
+ * @param product The product object from `context.storefront.query()` returned by a server loader.
+ * @param variants The available product variants for the product. This can be an array of variants, a promise that resolves to an array of variants, or an object with a `product` key that contains the variants.
+ * @returns A new product object where the `selectedVariant` property is set to the variant that matches the current URL search params. If no variant is found, the original product object is returned. The `isOptimistic` property is set to `true` if the `selectedVariant` has been optimistically changed.
+ */
+export function useOptimisticProduct<
+  ProductWithSelectedVariant = OptimisticProductInput,
+  Variants = OptimisticProductVariants,
+>(
+  product: ProductWithSelectedVariant,
+  variants: Variants,
+): OptimisticProduct<ProductWithSelectedVariant> {
+  const navigation = useNavigation();
+  const [resolvedVariants, setResolvedVariants] = useState<
+    Array<PartialDeep<ProductVariant>>
+  >([]);
+
+  useEffect(() => {
+    Promise.resolve(variants)
+      .then((productWithVariants) => {
+        if (productWithVariants) {
+          setResolvedVariants(
+            productWithVariants instanceof Array
+              ? productWithVariants
+              : (productWithVariants as unknown as ProductWithVariants).product
+                  .variants.nodes || [],
+          );
+        }
+      })
+      .catch((error) => {
+        reportError(
+          new Error(
+            '[h2:error:useOptimisticProduct] An error occurred while resolving the variants for the optimistic product hook.',
+            {
+              cause: error,
+            },
+          ),
+        );
+      });
+  }, [variants]);
+
+  if (navigation.state === 'loading') {
+    const queryParams = new URLSearchParams(navigation.location.search);
+    let reportedError = false;
+
+    // Find matching variant
+    const selectedVariant =
+      resolvedVariants.find((variant) => {
+        if (!variant.selectedOptions) {
+          if (!reportedError) {
+            reportedError = true;
+            reportError(
+              new Error(
+                '[h2:error:useOptimisticProduct] The optimistic product hook requires your product query to include variants with the selectedOptions field.',
+              ),
+            );
+          }
+          return false;
+        }
+
+        return variant.selectedOptions.every((option) => {
+          return queryParams.get(option.name) === option.value;
+        });
+      }) || (product as OptimisticProductInput).selectedVariant;
```

Comment:
Doesn't this have assumption that the dev created a query field named `selectedVariant`? Should this be documented?

```
const PRODUCT_FRAGMENT = `#graphql
  fragment Product on Product {
    id
    selectedVariant: variantBySelectedOptions(selectedOptions: $selectedOptions, ignoreUnknownOptions: true, caseInsensitiveMatch: true) {
      ...ProductVariant
    }
```

So ...

This will use `selectedVariant` if `resolvedVariants` is not available and if `selectedVariant` isn't available it just return the product.

If I'm getting this correctly, `selectedVariant` is serving as a quick select if it's available ... then how come it's not

```
const selectedVariant = (product as OptimisticProductInput).selectedVariant || resolvedVariants.find(...)
```

Reply:
Yes, that requirement might need to be more documented. The issue is `selectedVariant` might already be defined, because the user has selected a variant. But if they select a new variant, it should ignore what the current "selectedVariant` is, and calculate a new one based on the pending navigation. Otherwise it would only be "optimistic" for the first variant selection, not subsequent ones.